fix: #227 concurrent modification error

miquelbeltran opened the pull request:
## fix: #227 concurrent modification error

### Description :memo:

Fixes #227 

A concurrent modification was happening in the Breadcrumbs while the message serialization was running.

To avoid this issue, the following changes were made:

- Copy the list of breadcrumbs when building the Raygun message, rather than using the same list instance
- Copy as well the Tags and Custom Data payloads to avoid similar issues
- Ensure modifying breadcrumbs happens synchronously

**Type of change**

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Updates**

See description

### Screenshots :camera:



### Test plan :test_tube:

Tested with app currently in development

### Author to check :eyeglasses:
- [ ] Project and all contained modules build
- [ ] Tested on appropriate devices/emulators and SDK versions
- [ ] Reviewed by another developer
- [ ] Appropriate documentation written (code comments, internal docs)

### Reviewer to check :heavy_check_mark:
- [ ] Change has been tested on a device/emulator
- [ ] Code is written to standards
- [ ] Appropriate tests have been written (code comments, internal docs)

miquelbeltran committed:
fix: #227 synchronize breadcrumb array modification and use copies for msg serialization

miquelbeltran committed:
spotless

velocitysystems commented on provider/src/main/java/com/raygun/raygun4android/CrashReporting.kt:
```diff
@@ -48,14 +48,20 @@ object CrashReporting {
         recordBreadcrumb(breadcrumb)
     }
 
+    private val breadcrumbLock = Any()
```
Rather than a lock we could use `CopyOnWriteArrayList<RaygunBreadCrumbMessage>()`.
This is a thread-safe variant of `ArrayList<T>`.
- Reads are lock-free and always safe
- Writes just copy the list _internally_

TheRealAgentK replied:
I think both the `synchronized` block and the `CopyOnWriteArrayList` approach would work.

Using the COWAL seems a bit cleaner though.

miquelbeltran committed:
improve solution

miquelbeltran commented:
Did a couple of changes:
- Switched to `CopyOnWriteArrayList` as proposed
- Modified the `mergeList` and `mergeMap` methods so they always return a copy (not 100% related to this bug but should avoid future similar issues)
